### ISSUE TITLE
Bug on 3c33e15 with Board testing

### ISSUE THREAD

**Issue by Deophius** [Sudoku-dev/Sudoku#51]
## Description
When I build using `make`, it does build an exe. After that, when I run it directly, it just crashes. But when I load it into the `gdb` debugger, it worked quite well, reporting that no errors detected.

**Comment by Deophius** [comment]
# My First guess

Let's first look through the calling line.
`Board::Board()` -> `std::vector<Record::Cell>::resize(std::size_t)` -> `Cell::Cell()` -> `Cell::Cell(num, state)` -> `Cell::value(int num, State inState)`

In this line, the last function is suspicious. It will not work if the state is sure.
Before construction, there should be no states.
But the allocator will fill it with 0 first, and then pass it to the constructor.
When the constructor sees the object, its state property is set to 0, which is Sure.
So value() will not work if it is allocated on the heap, but will work if it's on the stack.
As for debugging, perhaps the debugger optimized some calls out, so the memory block for the cell is still random, making value work correctly.
In summary, the current implementation violated the rule "never read before initialize".

Only changing the value of Sure will work for a time, but not long.
There will be a time when the linker allocates a memory block with a "random" number that happened to be the new Sure value.
So we must prevent reading before initializing.
To do this, we can first set the properties in ctor, and then pass it on to `value()`.

Of course, this is only a guess.

**Comment by Deophius** [comment]
It seems that this can fix the issue.
See commit _889391f_, right after the commit _3c33e15_.
